src/main/java/org/broadinstitute/hellbender/tools/spark/validation/CompareDuplicates.java

```diff
@@ -0,0 +1,203 @@
+        JavaPairRDD<String, Tuple2<Iterable<GATKRead>, Iterable<GATKRead>>> cogroup = firstKeyed.cogroup(secondKeyed);
+
+        Broadcast<SAMFileHeader> broadcastHeader = ctx.broadcast(getHeaderForReads());
+
+        // Produces an RDD of MatchTypes, e.g., EQUAL, DIFFERENT_REPRESENTATIVE_READ, etc. per MarkDuplicates key,
+        // which is approximately start position x strand.
+        JavaRDD<Utils.MatchType> tagged = cogroup.map(v1 -> {
+            SAMFileHeader header = broadcastHeader.getValue();
+
+            Iterable<GATKRead> iFirstReads = v1._2()._1();
+            Iterable<GATKRead> iSecondReads = v1._2()._2();
+
+            return Utils.getDupes(iFirstReads, iSecondReads, header);
+        });
+
+        Map<Utils.MatchType, Integer> tagCountMap = tagged.mapToPair(v1 ->
+                new Tuple2<>(v1, 1)).reduceByKey((v1, v2) -> v1 + v2).collectAsMap();
+
+        if (tagCountMap.get(Utils.MatchType.SIZE_UNEQUAL) != null) {
+            throw new UserException("The number of reads by the MarkDuplicates key were unequal, indicating that the BAMs are not the same");
+        }
+        if (tagCountMap.get(Utils.MatchType.READ_MISMATCH) != null) {
+            throw new UserException("The reads grouped by the MarkDuplicates key were not the same, indicating that the BAMs are not the same");
+        }
+        if (printSummary) {
+            Utils.MatchType[] values = Utils.MatchType.values();
+            HashSet<Utils.MatchType> matchTypes = Sets.newHashSet(values);
+            System.out.println("##############################");
+            matchTypes.forEach(s -> System.out.println(s + ": " + tagCountMap.getOrDefault(s, 0)));
+        }
+
+        if (crashOnDiff) {
+            for (Utils.MatchType s : Utils.MatchType.values()) {
+                if (s != Utils.MatchType.EQUAL) {
+                    if (tagCountMap.get(s) != null)
+                    throw new UserException("found difference between the two BAMs: " + s + " with count " + tagCountMap.get(s));
+                }
+            }
+        }
+    }
+}
+
+
+class Utils implements Serializable {
+    private static final long serialVersionUID = 1L;
+
+    enum MatchType {
+        EQUAL,          // The same number of reads per key, same number of duplicates, and same representative read.
+        SIZE_UNEQUAL,   // Difference number of reads at this key (this means the BAMs are different).
+        READ_MISMATCH,  // The reads at this key are different for the two BAMs (which means the BAMs are different).
+        DIFF_NUM_DUPES, // The reads are the same from each BAM, but there are different number of duplicates marked.
+        DIFFERENT_REPRESENTATIVE_READ,  // The reads are the same and the same number of duplicates, but with a
+                                        // different representative read (usually this is from a tie-breaker).
+    }
+
+    static boolean isNonPrimary(GATKRead read) {
```

Comment:
this is a duplicate of MarkDuplicatesSpark.isNonPrimary - extract to a method on ReadUtils and call from both places